### PR TITLE
feat(inputs.kernel): Add Pressure Stall Information

### LINKED ISSUE #6760: Add input plugin for PSI (Pressure Stall Information)
## Feature Request

Facebook implemented PSI (Pressure Stall Information) for Linux 4.20 and higher. PSI presents high quality information about the pressure the IO, memory and CPU is under. More information here:

https://facebookmicrosites.github.io/psi/

### Proposal:
Input plugin for PSI
### Current behavior:
None
### Desired behavior:
Support/Input plugin for PSI 
### Use case: 
Good indicator if a system is overloaded

### PR DESCRIPTION
**Pull request by iBug**
## Summary

Implemented an input plugin for Pressure Stall Information (Linux v4.20+).

Part of the code was derived from [gridscale/linux-psi-telegraf-plugin](https://github.com/gridscale/linux-psi-telegraf-plugin), which is available under the MIT license.

I will be after this PR as long as I am around.

## Checklist

- [x] No AI generated code was used in this PR

## Related issues

Resolves #6760

### CODE REVIEW

**Comment by srebhan** [comment]
Thanks for your contribution @iBug! Before giving this a full review I would like to ask why you do not implement this feature as part of the [inputs.kernel plugin](https://github.com/influxdata/telegraf/tree/master/plugins/inputs/kernel)? You could easily add this as a new `collect` ID and make this explorable to all users of `kernel` which is probably the target audience anyway!?!? 

Furthermore, I suggest to merge the `pressure` and `pressureTotal` metrics to something like
```text
pressure,resource=cpu some_avg10=1.53,some_avg60=1.87,some_avg300=1.73,some_total=1088168194,full_avg10=0,full_avg60=0,full_avg300=0,full_total=0 1700000000000000000
pressure,resource=memory some_avg10=0,some_avg60=0,some_avg300=0,some_total=3463792,full_avg10=0,full_avg60=0,full_avg300=0,full_total=1429641 1700000000000000000
pressure,resource=io some_avg10=0,some_avg60=0,some_avg300=0,some_total=68568296,full_avg10=0,full_avg60=0,full_avg300=0,full_total=54982338 1700000000000000000
```

What do you think @iBug?

**Comment by iBug** [comment]
> implement this feature as part of the [inputs.kernel plugin](https://github.com/influxdata/telegraf/tree/master/plugins/inputs/kernel)

Very good suggestion. I didn't realize this was possible when I started this PR.

May I ask what should I do now? Start migrating this entire code to `inputs.kernel`, wait for more reviews / suggestions / discussions, or prepare a separate PR?

> merge the `pressure` and `pressureTotal` metrics

Merging `pressure` and `pressureTotal` makes total sense, but I'm quite hesitant on merging `some` and `full` tags. Specifically for `resource=cpu`, there is no `type=full`. Should I just omit the `full_*` values for `resource=cpu`?

**Comment by srebhan** [comment]
@iBug let me discuss this with the team...

**Comment by srebhan** [comment]
@iBug we discussed this PR yesterday and can live with the current metric format to keep the Gauge vs. Counter properties. However, we do also agree on merging this plugin into `inputs.kernel` as this is a natural place for this information. I don't think this is too hard as you basically need to move the contents of `psi_linux.go` there and guard the execution of the function with `ConfigCollect` containing  the feature flag (e.g. `"psi"`)... What do you think?

I can assist you in the merge if you want and permit pushes to your PR branch in your repository...

**Comment by iBug** [comment]
I just moved all the code into `inputs.kernel` and I think this PR is good to go.

> permit pushes to your PR branch

I forked the repository to an organization named `iBug-forks`, and it seems like "allow edits from maintainers" [is only for user-owned repos](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork). Meanwhile, GitHub [doesn't allow](https://stackoverflow.com/a/42381583/5958455) changing the incoming branch once a PR is created, so I'm afraid I'll have to commit & push all the code myself.

Fortunately I'll be able to spare some minutes daily in the upcoming days, so you can just comment and I'll handle the code. Alternatively I can close this PR, re-fork and re-submit another PR so you can push.

File: plugins/inputs/kernel/psi.go
Comment (srebhan) [inline plugins/inputs/kernel/psi.go:74]: Could we merge the two functions? Like
```suggestion
func (*Kernel) gatherPressure(acc telegraf.Accumulator) error {
	fs, err := procfs.NewDefaultFS()
	if err != nil {
		return fmt.Errorf("procfs not available: %w", err)
	}

	var metrics []telegraf.Metrics
	for _, resource := range []string{"cpu", "memory", "io"} {
		stats, err := fs.PSIStatsForResource(resource)
		if err != nil {
			return nil, fmt.Errorf("failed to read %s pressure: %w", resource, err)
		}

		// "Some" metrics total and average
		tags := map[string]string{
			"resource": resource,
			"type":     "some",
		}
		acc.AddGauge(
			"pressure",
			map[string]interface{}{
				"avg10":  stat.Some.Avg10,
				"avg60":  stat.Some.Avg60,
				"avg300": stat.Some.Avg300,
			},
			tags,
		)
		acc.AddCounter(
			"pressure",
			map[string]interface{}{
				"total": stats.Some.Total,
			},
			tags,
		)
		
		// "Full" metrics total and average
		tags = map[string]string{
			"resource": resource,
			"type":     "full",
		}
		acc.AddGauge(
			"pressure",
			map[string]interface{}{
				"avg10":  stat.Full.Avg10,
				"avg60":  stat.Full.Avg60,
				"avg300": stat.Full.Avg300,
			},
			tags,
		)
		// The total metrics is only conditionally added for "cpu" as this is all-empty and will be omitted.
		if resource != "cpu" {
			acc.AddCounter(
				"pressure",
				map[string]interface{}{
					"total": stats.Full.Total,
				},
				tags,
			)
		}
	}

	return nil
}
```

**Review by srebhan — COMMENTED** [review]
Thanks for merging this to the `kernel` plugin @iBug! Two comments from my side...

File: plugins/inputs/kernel/psi.go
Comment (iBug) [reply plugins/inputs/kernel/psi.go:74]: Merging these functions sound like a good idea, but how could I implement unit tests? Since I'm relying on Prometheus to both read and parse `/proc/pressure/*`, there's pretty much nothing to test after merging collection and submission.

Comment (srebhan) [reply plugins/inputs/kernel/psi.go:74]: It's probably a good idea to honor the `$HOST_PROC` environment variable like [gopsutil does](https://github.com/shirou/gopsutil?tab=readme-ov-file#usage) to allow running telegraf in a container which has the host's `/proc` mounted. This also has the great advantage to be able to fake the `/proc` dir for tests by placing the corresponding file-structure it in a `testdata` directory and set that environment variable accordingly...

Comment (srebhan) [reply plugins/inputs/kernel/psi.go:74]: Oh I saw you already do have `psiDir` to specify where to find the stuff. So how about using `filepath.Dirname` on that path and then use [procfs's NewFS()](https://pkg.go.dev/github.com/prometheus/procfs#NewFS) instead of `procfs.NewDefaultFS()`?

Comment (iBug) [reply plugins/inputs/kernel/psi.go:74]: You reminded me of using [`procfs.NewFS`](https://pkg.go.dev/github.com/prometheus/procfs#NewFS) instead of `NewDefaultFS`, which is *the* solution we need.

Comment (iBug) [reply plugins/inputs/kernel/psi.go]: I think this is done now.

File: plugins/inputs/kernel/psi_test.go
Comment (srebhan) [inline plugins/inputs/kernel/psi_test.go:75]: Nowadays we rather use ´testutil.RequireMetricsEqual` to do the testing as it makes it easier to see what is actually expected. And I would spell out those metrics here instead of a loop to directly see the expectation... So how about

```suggestion
	k := Kernel{
		psiDir:        "testdata/pressure",
		ConfigCollect: []string{"psi"},
	}
	require.NoError(t, k.Init())

	var acc testutil.Accumulator
	require.NoError(t, k.gatherPressure(&acc))

	expected := []telegraf.Metric{
		metric.New(
			"pressure",
			map[string]string{
				"resource": "cpu",
				"type":     "some",
			},
			map[string]interface{}{
				"avg10":  float64(10),
				"avg60":  float64(60),
				"avg300": float64(300),
			},
			time.Unix(0,0),
			telegraf.Gauge,
		),
		metric.New(
			"pressure",
			map[string]string{
				"resource": "cpu",
				"type":     "some",
			},
			map[string]interface{}{
				"total": uint64(114514),
			},
			time.Unix(0,0),
			telegraf.Counter,
		),
		metric.New(
			"pressure",
			map[string]string{
				"resource": "memory",
				"type":     "some",
			},
			map[string]interface{}{
				"avg10":  float64(10),
				"avg60":  float64(60),
				"avg300": float64(300),
			},
			time.Unix(0,0),
			telegraf.Gauge,
		),
		metric.New(
			"pressure",
			map[string]string{
				"resource": "memory",
				"type":     "some",
			},
			map[string]interface{}{
				"total": uint64(114514),
			},
			time.Unix(0,0),
			telegraf.Counter,
		),
		metric.New(
			"pressure",
			map[string]string{
				"resource": "io",
				"type":     "some",
			},
			map[string]interface{}{
				"avg10":  float64(10),
				"avg60":  float64(60),
				"avg300": float64(300),
			},
			time.Unix(0,0),
			telegraf.Gauge,
		),
		metric.New(
			"pressure",
			map[string]string{
				"resource": "io",
				"type":     "some",
			},
			map[string]interface{}{
				"total": uint64(114514),
			},
			time.Unix(0,0),
			telegraf.Counter,
		),
		metric.New(
			"pressure",
			map[string]string{
				"resource": "memory",
				"type":     "full",
			},
			map[string]interface{}{
				"avg10":  float64(1),
				"avg60":  float64(6),
				"avg300": float64(30),
			},
			time.Unix(0,0),
			telegraf.Gauge,
		),
		metric.New(
			"pressure",
			map[string]string{
				"resource": "memory",
				"type":     "full",
			},
			map[string]interface{}{
				"total": uint64(11451),
			},
			time.Unix(0,0),
			telegraf.Counter,
		),
		metric.New(
			"pressure",
			map[string]string{
				"resource": "io",
				"type":     "full",
			},
			map[string]interface{}{
				"avg10":  float64(1),
				"avg60":  float64(6),
				"avg300": float64(30),
			},
			time.Unix(0,0),
			telegraf.Gauge,
		),
		metric.New(
			"pressure",
			map[string]string{
				"resource": "io",
				"type":     "full",
			},
			map[string]interface{}{
				"total": uint64(11451),
			},
			time.Unix(0,0),
			telegraf.Counter,
		),
	}
	
	actual := acc.GetTelegrafMetrics()
	testutil.RequireMeticsEqual(t, expected, actual, testutil.IgnoreTime(), testutil.SortMetrics())
```

File: plugins/inputs/kernel/kernel.go
Comment (srebhan) [inline plugins/inputs/kernel/kernel.go:161]: Let's make the error specific here to easily find the spot on debugging...
```suggestion
			return fmt.Errorf("gathering pressure failed: %w", err)
```

File: plugins/inputs/kernel/kernel.go
Comment (srebhan) [inline plugins/inputs/kernel/kernel.go:16]: We usually have 3 groups of imports, builtins, 3rd-party libs and telegraf internal ones (in this order), each ordered alphabetically within the group. This make spotting external dependencies easier... So please do
```suggestion
	"github.com/prometheus/procfs"
	
	"github.com/influxdata/telegraf"
	"github.com/influxdata/telegraf/plugins/inputs"
```
Same for the `psi.go` and test files.

**Review by srebhan — COMMENTED** [review]
Very nice @iBug, it's really a joy to work with you! I would like to convert the test to use `testutil.RequireMetricsEqual` as this prints a nice diff on fail and allows to specify the expected metrics as such. The other two comments are minor...

We are almost there... :-)

File: plugins/inputs/kernel/kernel.go
Comment (iBug) [reply plugins/inputs/kernel/kernel.go:16]: I use `goimports` as a replacement for `gofmt`, which only arranges imports into stdlib and non-stdlib. I also have automatic `goimports` on save configured for all my editors. Not sure how to easily get over this one.

Comment (iBug) [reply plugins/inputs/kernel/kernel.go]: Looks like there's a `-local` flag that needs some configuration (it never had an "automatically detect from `go.mod`" feature). I ran it on `*.go` files.

File: plugins/inputs/kernel/kernel.go
Comment (iBug) [reply plugins/inputs/kernel/kernel.go:162]: I've reverted this suggestion, since there's only one routine in `k.gatherPressure` that [returns an already-formatted error](https://github.com/influxdata/telegraf/pull/14507/files#diff-ae8b99a6c960a65f7742f67c1270ddc6cc1c967c038562282c1ee31be86f938aR20). Doesn't seem useful to me to wrap it up again with more verbosity.

**Review by powersj — APPROVED** [review]
Thank you very much for the back and forth on this PR!

**Review by srebhan — APPROVED** [review]
Thanks for all you work @iBug! It was fun to work with you! Looking forward to your next PR. :-)

**Comment by ajfriesen** [comment]
Thank you very much for integrating the psi plugin we created.

Very much appreciated 🙏

**Pull request merged**
